### ISSUE TITLE
[RFC] Default torch.compile backend customization

### ISSUE THREAD

**Issue by bsochack** [pytorch/pytorch#136118]
### 🚀 The feature, motivation and pitch

This RFC proposes an enhancement to torch.compile to improve its backend agnosticism. The goal is to enable a more seamless experience for users working with devices that may not be well-supported by the current default backend, Inductor. By allowing for the customization of the default backend, we aim to improve the overall usability of torch.compile across a wider range of hardware.

**Background:** torch.compile currently defaults to the Inductor backend, which is optimized for GPUs and devices that benefit from Inductor's loop-level IR. Users can select the Inductor backend explicitly by specifying backend="Inductor" or implicitly by omitting the backend parameter. However, this can lead to compilation issues on devices without Inductor integration.

**Problem Statement:** The implicit selection of the Inductor backend can cause compatibility issues for devices that do not support it. While explicit backend selection by the user should be respected, the implicit case presents an opportunity to adapt to different device types.

**Proposal:** We propose to introduce a mechanism that allows for the customization of the default backend used by torch.compile. This would enable device-specific backends to be set as the default, thereby making user code more robust and backend-agnostic. For example, on an HPU, the default backend could be set to "hpu_backend," eliminating the need for code modifications when no backend is explicitly defined.

**Implementation:** The customization of the default backend could be achieved through a new configuration option within torch._dynamo.config or a similar in a dedicated configuration space. Device extensions could then specify their preferred backend during their initialization phase, such as during the autoload process (as discussed in https://github.com/pytorch/pytorch/issues/122468).

@albanD @ezyang @jgong5 please provide feedback.

**Comment by ezyang** [comment]
The general concept seems fine, not sure about the details

**Comment by ezyang** [comment]
cc @jansel

**Comment by jansel** [comment]
I think using a global config flag would be bad, since changing that global option might break things like CPU and (unless the user changed it themselves) might lead to confusing behavior.  If there are a way to bind it to a specific device that would be better.

Though I'm not sure how big a problem this is.  For a custom device the user would already need to change their code to do `.to(custom_device)`, so picking a backend doesn't seem like that much extra work.

**Comment by bsochack** [comment]
@ezyang, @jansel, thank you for comments.

It seems like a global config is not the right approach. I will try to propose a different solution with the purpose of customization only during device extension initialization.

> Though I'm not sure how big a problem this is. For a custom device the user would already need to change their code to do .to(custom_device), so picking a backend doesn't seem like that much extra work.
I think we are facing 2 issues here:
- we see customers who don't want to change any line of code - zero line migration
- torch.compile can be used in 3rd party code/libraries where this customization can be a bit harder to make

This is true that `.to(custom_device)`  still need to be changed today. One possible solution is to introduce a concept of default device which can point to cuda, cpu, hpu or other device depends on platform, so the code could be even more device agnostic. This topic is probably for another discussion. I'm still not clear if it is a good idea to propose something like that.

**Comment by bsochack** [comment]
I did not reply to this:
>  since changing that global option might break things like CPU and (unless the user changed it themselves) might lead to confusing behavior

All torch.compile backends need to take care of CPU one way or another. FX graphs are not partitioned based on a device before they are passed to a backend. So all backends need to handle it.
There are at least 2 options here:
1. change it globally and a backend needs to decide how to handle CPU (it is already the case) i.e. when all tensors are on CPU, it can fallback to Inductor
2. check if any of tensors is custom device and only then change default behavior. There is some cost at runtime of it.

Do you see any cases when it is desired to mix devices and there is performance benefit of it?

**Comment by ezyang** [comment]
@eellison We're planning on partitioning for CUDA graphs right? Could we potentially extend this to device type partitioning (since getting CPU compute out of CUDA graph region is also in scope for that)

**Comment by jansel** [comment]
Let's start with a better error message when inductor doesn't support a device.  That should be an easy fix.

In the general case each graph has a set of used devices.  If that set is size 1 we could do something, though that doesn't feel that satifying.  If we add a parititoner we could have inductor delegate a subgraph.

**Comment by bsochack** [comment]
Ok, just to summarize, your vision is that a FX graph first will get into Inductor, Inductor before lowering to Loop-level IR (need to decide an exact point) will partition the graph between the devices and will provide option to pass partitions to dedicated backends.
So I assume there will be some method to customize device to backend mapping in Inductor. Is that correct?
i.e. for hpu device, we will pass partition to hpu_backend.

**Comment by jansel** [comment]
Yes, I think something like that would work.  Though it would only apply for devices inductor doesn't know how to compile.  Inductor does some cross-device optimizations today (e.g. moving `torch.randn(100).to(device)` from CPU to `device`).  You caan also register a custom loop-level IR backend for inductor to support new devices.

**Comment by bsochack** [comment]
Will it allow to specify decompositions per device ? does it mean that devices must align the ones registered by Inductor ?

**Comment by jansel** [comment]
Implemented in the most straightforward way (post grad), you could apply additional decomps but not remove decomps already applied.
